ide-plugin/src/org/jetbrains/plugins/ruby/ruby/codeInsight/types/RubyReturnTypeProvider.kt

```diff
@@ -37,6 +48,36 @@ class RubyReturnTypeProvider : AbstractRubyTypeProvider() {
             }
             return originType
         }
+        if (expr is RIdentifier && expr.isParameter) {
+            val method = expr.parentContainer as? RMethod ?: return null
+            val moduleName = (method.parentContainer as? RElementWithFQN)?.fqn?.fullPath ?: return null
+            val indexOfArgument = method.arguments.indexOfFirst { it.identifier == expr }
+
+            val info = MethodInfo(ClassInfo(moduleName), method.fqn.shortName, RVisibility.PUBLIC)
+
+            val callInfos: List<CallInfo> = RSignatureProviderImpl().getRegisteredCallInfos(info)
+
+            return callInfos.map {
+                if (indexOfArgument in 0 until it.argumentsTypes.size) {
+                    RTypeFactory.createTypeClassName(it.argumentsTypes[indexOfArgument], expr)
+                } else {
+                    REmptyType.INSTANCE
+                }
+            }.filter { it != REmptyType.INSTANCE }.distinctBy { it.name }.unionTypes()
+
+        }
         return null
     }
+}
+
+private fun List<RType>.unionTypes(startIndex: Int = 0, endIndex: Int = size - 1): RType {
```

Comment:
why do you need endIndex here?

Reply:
BTW, this method impl could be replaced by `List.reduce` :-)

Reply:
`List.reduce` won't help here as it accumulates values from left to right but I need to maximize tree height